src/Categories/Double/Core.agda

```diff
@@ -0,0 +1,297 @@
+{-# OPTIONS --without-K --safe #-}
+module Categories.Double.Core where
+
+open import Level
+
+open import Relation.Binary using (Rel; IsEquivalence; Setoid; REL)
+open import Relation.Binary.PropositionalEquality as ≡ using (_≡_)
+import Relation.Binary.Reasoning.Setoid as SetoidR
+
+import Categories.Category.Unbundled as Cat
+
+-- Equality of squares in a double category
+module _ {o ℓ e : Level} {Obj : Set o} (Hor Ver : Cat.Category Obj ℓ e) where
+  private
+    module H = Cat.Category Hor
+    module V = Cat.Category Ver
+    _≈ₕ_ : ∀ {A B} → Rel (A H.⇒ B) e
+    _≈ₕ_ = H._≈_
+    _≈ᵥ_ : ∀ {A B} → Rel (A V.⇒ B) e
+    _≈ᵥ_ = V._≈_
+  record SqEquality
+    {T₁ T₂ B₁ B₂ : Obj}
+    (hT₁ hT₂ : T₁ H.⇒ T₂)
+    (hB₁ hB₂ : B₁ H.⇒ B₂)
+    (vL₁ vL₂ : T₁ V.⇒ B₁)
+    (vR₁ vR₂ : T₂ V.⇒ B₂) : Set e where
+      field
+        horT≈ : hT₁ ≈ₕ hT₂
+        horB≈ : hB₁ ≈ₕ hB₂
+        verL≈ : vL₁ ≈ᵥ vL₂
+        verR≈ : vR₁ ≈ᵥ vR₂
+
+module _ {o ℓ e : Level} {Obj : Set o} (Hor Ver : Cat.Category Obj ℓ e) where
+  private
+    module H = Cat.Category Hor
+    module V = Cat.Category Ver
+  dual≈ :
+    ∀ {T₁ T₂ B₁ B₂ : Obj} →
+    {hT₁ hT₂ : T₁ H.⇒ T₂} →
+    {hB₁ hB₂ : B₁ H.⇒ B₂} →
+    {vL₁ vL₂ : T₁ V.⇒ B₁} →
+    {vR₁ vR₂ : T₂ V.⇒ B₂} →
+    SqEquality Hor Ver hT₁ hT₂ hB₁ hB₂ vL₁ vL₂ vR₁ vR₂ →
+    SqEquality Ver Hor vL₁ vL₂ vR₁ vR₂ hT₁ hT₂ hB₁ hB₂
+  dual≈ S = record
+      { horT≈ = S.verL≈
+      ; horB≈ = S.verR≈
+      ; verL≈ = S.horT≈
+      ; verR≈ = S.horB≈
+      }
+    where module S = SqEquality S
+
+-- Basic definition of a |Double Category|
```

Comment:
It would be good to document somewhere (probably the top) what level of strictness you are using. It looks like these are "maximally" strict, i.e., equivalent to strict categories internal to cat. But often people say double category to mean vertically strict and horizontally weakly associative/unital, so best to clarify that up front.

Reply:
I think this might just be because it's WIP. Already the definition is weaker than usual because of the concept of equality on squares being used is much less strict.

Reply:
Not completely sure what would be best to put here. This feels like a correct definition for a setoid enriched category in the sense that the source and target for 2-cells must be equipped with maps between the associated hom-setoids and so equality of 2-cells must include the data of equality between each of the 1-cells. What do you both think would be best?

Reply:
I agree that this is the most 'natural' definition of Double Category (and agrees with the one on nLab) for E-Categories. There are indeed different definitions for Pseudo Double and Double Bicategory.